What's new in .NET Aspire 9.2 (#2877)
dotnet/docs-aspire: docs/whats-new/dotnet-aspire-9.2.md

    ```

    > [!IMPORTANT]
    > The `dotnet new update` command updates all of your templates to the latest version.

If your app host project file doesn't have the `Aspire.AppHost.Sdk` reference, you might still be using .NET Aspire 8. To upgrade to 9.0, follow [the upgrade guide](../get-started/upgrade-to-aspire-9.md).

## 🚧 App host project file changes

The .NET Aspire app host project file no longer requires the `IsAspireHost` property. This property was moved to the `Aspire.AppHost.Sdk` SDK, therefore, you can remove it from your project file. For more information, see [dotnet/aspire issue #8144](https://github.com/dotnet/aspire/pull/8144).

<!--
## 📈 Dashboard usage telemetry

.NET Aspire collects usage telemetry from the dashboard by default. This telemetry helps us understand how you use the dashboard and what features are most important to you. We use this information to prioritize our work and improve the dashboard experience. You can opt out of this telemetry by setting the `DOTNET_DASHBOARD_ENABLE_TELEMETRY` environment variable to `false`. For more information, see [.NET Aspire dashboard usage telemetry](../fundamentals/dashboard/usage-telemetry.md).

> -->

Reviewer: Can we just remove this now it got punted to 9.3?

Author: Yeah, I had it commented out for that reason - it wouldn't render anyway, but I will remove it.

> ## 🚫 Disable dashboard resource graphs

Reviewer 2: I feel like this is something strange to promote some prominently. Maybe we should have an errata section in the bottom for this?

Reviewer 3: +1. This was just added for folks who are hosting the dashboard, have resources, but don't have relationships, so the graph doesn't add value.

The average user won't use this and won't care that it's there.

## App Host

### 🚧 project file changes

The .NET Aspire app host project file no longer requires the `IsAspireHost` property. This property was moved to the `Aspire.AppHost.Sdk` SDK, therefore, you can remove it from your project file. For more information, see [dotnet/aspire issue #8144](https://github.com/dotnet/aspire/pull/8144).

### Custom Resource URLs

Resources can now define custom URLs! This makes it easier to build custom experiences for your resources. For example, you can define a custom URL for a database resource that points to the database management console. This makes it easier to access the management console directly from the dashboard, you can even give it a friendly name.

```C#
var catalogDb = builder.AddPostgres("postgres")
                    .WithDataVolume()
                    .WithPgAdmin(resource =>
                    {
                        resource.WithUrlForEndpoint("http", u => u.DisplayText = "PG Admin");
                    })
                    .AddDatabase("catalogdb");
```

The above code sets the display text for the `PG Admin` URL to `PG Admin`. This makes it easier to access the management console directly from the dashboard.

> 

Reviewer 3: ```suggestion

## 🔧 Dashboard

.NET Aspire 9.2 adds new features to the [dashboard](/docs/fundamentals/dashboard/overview.md), making it a more powerful developer tool than ever. The following features were added to the dashboard in .NET Aspire 9.2:

### 🧩 Resource graph

The resource graph is a new way to visualize the resources in your apps. It displays a graph of resources, linked by relationships. Click the 'Graph' tab on the Resources page to view the resource graph.

:::image type="content" source="media/dashboard-resource-graph.gif" lightbox="media/dashboard-resource-graph.gif" alt-text="Recording of using the dashboard resource graph.":::

### 🎨 Resources icons

We've added resource icons to the resources page! The icon color matches the resource's telemetry in structured logs and traces.

:::image type="content" source="media/dashboard-resource-icons.png" lightbox="media/dashboard-resource-icons.png" alt-text="Screenshot of dashboard resource's page showing the new resource icons.":::

### ⏯️ Pause and resume telemetry

New buttons were added to the **Console logs**, **Structured logs**, **Traces** and **Metrics** pages to pause collecting telemetry. Click the pause button again to resume collecting telemetry.

This feature allows you to pause telemetry in the dashboard while continuing to interact with your app.

:::image type="content" source="media/dashboard-remove-telemetry.png" lightbox="media/dashboard-remove-telemetry.png" alt-text="Screenshot of the dashboard showing the pause button.":::

### ❤️‍🩹 Metrics health warning

The dashboard now warns you when a metric exceeds the configured cardinality limit. Once exceeded, the metric no longer provides accurate information.

:::image type="content" source="media/dashboard-cardinality-limit.png" lightbox="media/dashboard-cardinality-limit.png" alt-text="Screenshot of a metric with the cardinality limit warning.":::

### 🕰️ Console logs UTC

Console logs now supports UTC timestamps. The setting is accessible via the console logs options button.

:::image type="content" source="media/dashboard-console-logs-utc.png" lightbox="media/dashboard-console-logs-utc.png" alt-text="Screenshot of console logs page showing the UTC timestamps option.":::

### 🔎 Trace details search text box

We've added a search text box to trace details. Now you can quickly filter large traces to find the exact span you need.

```

Reviewer 3: Pngs and gifs were added to a teams conversation. Or you can get them from my social media.

Author: Thanks @JamesNK - I'll get these updates in.

### WithHttpCommand

Adds WithHttpCommand(), which lets you define a resource command that sends an HTTP request to your app during development. Useful for triggering endpoints like seed or reset from the dashboard.

> ```c#    

Reviewer 4: ```suggestion
```c#
```

if (builder.Environment.IsDevelopment())
{
    var resetDbKey = Guid.NewGuid().ToString();
    catalogDbApp.WithEnvironment("DatabaseResetKey", resetDbKey)
                .WithHttpCommand("/reset-db", "Reset Database",
                    commandOptions: new()
                    {
                        Description = "Reset the catalog database to its initial state. This will delete and recreate the database.",
                        ConfirmationMessage = "Are you sure you want to reset the catalog database?",
                        IconName = "DatabaseLightning",
                        PrepareRequest = requestContext =>
                        {
                            requestContext.Request.Headers.Add("Authorization", $"Key {resetDbKey}");
                            return Task.CompletedTask;
                        }
                    });
}
```

### ConnectionStringResource

We've introduced a new ConnectionStringResource type that makes it easier to build dynamic connection strings without defining a separate resource
type. This makes it easier to work with build dynamic parameterized connection strings.

   ```c#
   var apiKey = builder.AddParameter("apiKey");

>    var cs = builder.ApplicationBuilder.AddConnectionString("openai", csb =>

Reviewer 2: Should probably get rid of `ApplicationBuilder` here. Assume this is just copy pasta.

```c#
var vault = builder.AddAzureKeyVault("kv");

var redis = builder.AddAzureRedis("redis")
                   .WithAccessKeyAuthentication(vault);

builder.AddProject<Projects.Api>("api")
       .WithReference(redis);
```

**Previous behavior:**  
`azd` created and managed secrets using Bicep, with no visibility in the app model. Secrets were handled implicitly and couldn’t be customized in C#.

**New behavior in 9.2:**  
Calling `WithKeyAccessAuthentication` or `WithPasswordAuthentication` now creates an actual `AzureKeyVaultResource` (or accepts a reference to one), and stores connection strings there. Secret names follow the pattern `connectionstrings--{resourcename}` to prevent naming conflicts with other vault entries.

### Improved Default Permissions for Key Vault References

When referencing a Key Vault, Aspire previously granted the broad **Key Vault Administrator** role by default. In 9.2, this has been changed to **Key Vault Secrets User**, which provides read-only access to secrets—suitable for most application scenarios.

This update continues the security-focused improvements in this release.

## Deployment

> Here’s the updated section with your additions, keeping the tone and structure consistent:

Reviewer 2: LLM bleed

1. Update to the latest [.NET Aspire templates](../fundamentals/aspire-sdk-templates.md) by running the following .NET command line:

    ```dotnetcli
    dotnet new update
    ```

    > [!IMPORTANT]
    > The `dotnet new update` command updates all of your templates to the latest version.

If your app host project file doesn't have the `Aspire.AppHost.Sdk` reference, you might still be using .NET Aspire 8. To upgrade to 9.0, follow [the upgrade guide](../get-started/upgrade-to-aspire-9.md).

## 🖥️ App Host

The [app host](../fundamentals/app-host-overview.md) is the core of .NET Aspire, providing the local hosting environment for your distributed applications. In .NET Aspire 9.2, we've made several improvements to the app host:

### 🚧 project file changes

The .NET Aspire app host project file no longer requires the `IsAspireHost` property. This property was moved to the `Aspire.AppHost.Sdk` SDK, therefore, you can remove it from your project file. For more information, see [dotnet/aspire issue #8144](https://github.com/dotnet/aspire/pull/8144).

> ### 🔗 Define custom resource URLs

Reviewer 4: @DamianEdwards

Adds WithHttpCommand(), which lets you define a resource command that sends an HTTP request to your app during development. Useful for triggering endpoints like seed or reset from the dashboard.

```csharp
if (builder.Environment.IsDevelopment())
{
    var resetDbKey = Guid.NewGuid().ToString();

    catalogDbApp.WithEnvironment("DatabaseResetKey", resetDbKey)
                .WithHttpCommand("/reset-db", "Reset Database",
                    commandOptions: new()
                    {
                        Description = "Reset the catalog database to its initial state. This will delete and recreate the database.",
                        ConfirmationMessage = "Are you sure you want to reset the catalog database?",
                        IconName = "DatabaseLightning",
                        PrepareRequest = requestContext =>
                        {
                            requestContext.Request.Headers.Add("Authorization", $"Key {resetDbKey}");
                            return Task.CompletedTask;
                        }
                    });
}
```

For more information, see [Custom HTTP commands in .NET Aspire](../fundamentals/http-commands.md).

> 
> 

Reviewer 4: ```suggestion

```

                .WithRedisInsight();
```

The `ImagePullPolicy` is set to `Always`, which means the image will always be pulled when the resource is created. This is useful for resources that are updated frequently.

### 📂 New container files API

In .NET Aspire 9.2, we've added a new `WithContainerFiles` API, a way to create files and folders inside a container at runtime by defining them in code. Under the hood, it uses `docker cp` / `podman cp` to copy the files in. Supports setting contents, permissions, and ownership—no bind mounts or temp files needed.

## 🤝 Integrations updates

Integrations are a key part of .NET Aspire, allowing you to easily add and configure services in your app. In .NET Aspire 9.2, we've made several updates to integrations:

### 🔐 Redis/Valkey/Garnet: Password support enabled by default

The Redis, Valkey, and Garnet containers enable password authentication by default. This is part of our goal to be secure by default—protecting development environments with sensible defaults while still making them easy to configure. Passwords can be set explicitly or generated automatically if not provided.

> ### Automatic database creation support for Postgres and SQL Server

Reviewer 4: @sebastienros

> ### Automatic database creation support for Postgres and SQL Server

Author: ```suggestion
### 💾 Automatic database creation support for Postgres and SQL Server
```

There's [plenty of feedback and confusion](https://github.com/dotnet/aspire/issues/7101) around the `AddDatabase` API. The name implies that it adds a database, but it didn't actually create the database. In .NET Aspire 9.2, the `AddDatabase` API now creates a database for the following hosting integrations:

| Hosting integration | API reference |
|--|--|
| [📦 Aspire.Hosting.SqlServer](https://www.nuget.org/packages/Aspire.Hosting.SqlServer) | <xref:Aspire.Hosting.SqlServerBuilderExtensions.AddDatabase*> |
| [📦 Aspire.Hosting.PostgreSql](https://www.nuget.org/packages/Aspire.Hosting.PostgreSql) | <xref:Aspire.Hosting.PostgresBuilderExtensions.AddDatabase*> |

> The Azure SQL and Azure PostgreSQL hosting integrations both expose an `AddDatabase` API, but they don't create a database—unless you call their respective `RunAsContainer` methods. For more information, see [Understand Azure integration APIs](../azure/integrations-overview.md#understand-azure-integration-apis).

Reviewer 5: @davidfowl @sebastienros - what does this line mean exactly?

Azure SQL And Azure PostgreSQL will create the database in Azure via bicep.

Reviewer 6: Correct, the database is created. The script won't run though so if it has more than CREATE DATABASE people might get confused.

We could imagine implementing it as part of a deployment script too to seed even and Azure database when it's deployed. Migration scripts.

Reviewer 5: I tweaked the wording here in https://github.com/dotnet/docs-aspire/pull/3066. If it needs to be adjusted, can you create a PR to change it?

By default, .NET Aspire will create an empty database if it doesn't exist. You can also optionally provide a custom script to run during creation for advanced setup or seeding.

Example using Postgres:

```csharp
var builder = DistributedApplication.CreateBuilder(args);

var postgres = builder.AddPostgres("pg1");

postgres.AddDatabase("todoapp")
    .WithCreationScript($$"""
        CREATE DATABASE {{databaseName}}
            ENCODING = 'UTF8';
        """);
```

For more information and examples of using the `AddDatabase` API, see:

- [Add PostgreSQL resource with database scripts](../database/postgresql-integration.md#add-postgresql-resource-with-database-scripts)
- [Add SQL Server resource with database scripts](../database/sql-server-integration.md#add-sql-server-resource-with-database-scripts)

The following hosting integrations don't currently support database creation:

- [📦 Aspire.Hosting.MongoDb](https://www.nuget.org/packages/Aspire.Hosting.MongoDb)
- [📦 Aspire.Hosting.MySql](https://www.nuget.org/packages/Aspire.Hosting.MySql)
- [📦 Aspire.Hosting.Oracle](https://www.nuget.org/packages/Aspire.Hosting.Oracle)

## ☁️ Azure integration updates

> In .NET Aspire 9.2, we've made several updates to Azure integrations, including:

Reviewer 4: ```suggestion
In .NET Aspire 9.2, we've made significant updates to Azure integrations, including:
```

- It enforces the principle of least privilege.
- It provides better isolation between apps in multi-service environments.

By assigning identities individually, Aspire can now scope role assignments more precisely—improving security, auditability, and alignment with Azure best practices.

This is a **behavioral breaking change** and may impact apps using:

- **Azure SQL Server** - Azure SQL only supports one Azure AD admin. With multiple identities, only the *last deployed app* will be granted admin access by default. Other apps will need explicit users and role assignments.

- **Azure PostgreSQL** - The app that creates the database becomes the owner. Other apps (like those running migrations or performing data operations) will need explicit `GRANT` permissions to access the database correctly.

See the [breaking changes](#-breaking-changes) section for more details.

This new identity model is an important step toward more secure and maintainable applications in Aspire. While it introduces some setup considerations, especially for database integrations, it lays the groundwork for better default security across the board.

> ### 🔑 WithRoleAssignments

Author: ```suggestion
### 🔑 Least-privilege role assignment functionality
```

                   .WithAccessKeyAuthentication(vault);

builder.AddProject<Projects.Api>("api")
       .WithReference(redis);
```

**Previous behavior:**

  `azd` created and managed secrets using Bicep, with no visibility in the app model. Secrets were handled implicitly and couldn't be customized in C#.

**New behavior in 9.2:**

  Calling `WithKeyAccessAuthentication` or `WithPasswordAuthentication` now creates an actual `AzureKeyVaultResource` (or accepts a reference to one), and stores connection strings there. Secret names follow the pattern `connectionstrings--{resourcename}` to prevent naming conflicts with other vault entries.

### 🔒 Improved default permissions for Azure Key Vault references

When referencing a Key Vault, Aspire previously granted the broad **Key Vault Administrator** role by default. In 9.2, this has been changed to **Key Vault Secrets User**, which provides read-only access to secrets—suitable for most application scenarios.

This update continues the security-focused improvements in this release.

## 🚀 Deployment improvements

We're excited to announce several new deployment features in .NET Aspire 9.2, including:

> ### 📦 Built-in publishers (Preview)

Reviewer 4: ```suggestion
### 📦 Publishers (Preview)
```

> Publishing integrations (or simply "publishers") are a new extensibility point in .NET Aspire that allow you to define how your distributed application gets transformed into deployable assets. Rather than relying on an [intermediate manifest format](../deployment/manifest-format.md), publishers can now plug directly into the application model to generate Docker Compose files, Kubernetes manifests, Azure resources, or whatever else your environment needs.

Reviewer 4: ```suggestion
Publishers are a new extensibility point in .NET Aspire that allow you to define how your distributed application gets transformed into deployable assets. Rather than relying on an [intermediate manifest format](../deployment/manifest-format.md), publishers can now plug directly into the application model to generate Docker Compose files, Kubernetes manifests, Azure resources, or whatever else your environment needs.
```

> When .NET Aspire launched, it introduced a deployment manifest format—a serialized snapshot of the application model. While useful for generating deployment targets like Bicep or Terraform, it burdened deployment tools with interpreting the manifest and resource authors with ensuring accurate serialization. This approach also complicated schema evolution and target-specific behaviors.

Reviewer 4: ```suggestion
When .NET Aspire launched, it introduced a deployment manifest format—a serialized snapshot of the application model. While useful it burdened deployment tools with interpreting the manifest and resource authors with ensuring accurate serialization. This approach also complicated schema evolution and target-specific behaviors.
```

Publishers simplify this process by working directly with the full application model in-process, enabling richer, more flexible, and maintainable publishing experiences.

The following publishers are available:

- [📦 Aspire.Hosting.Azure](https://www.nuget.org/packages/Aspire.Hosting.Azure)
- [📦 Aspire.Hosting.Docker (Preview)](https://www.nuget.org/packages/Aspire.Hosting.Docker)
- [📦 Aspire.Hosting.Kubernetes (Preview)](https://www.nuget.org/packages/Aspire.Hosting.Kubernetes)

> [!IMPORTANT]

> > The Docker and Kubernetes publishers were initially written by community contributor, [Dave Sekula](https://github.com/Prom3theu5)—a great example of the community stepping up to extend the model. 💜 Thank you, Dave!

Reviewer 4: ```suggestion
> The Docker and Kubernetes publishers were contributed by community contributor, [Dave Sekula](https://github.com/Prom3theu5)—a great example of the community stepping up to extend the model. 💜 Thank you, Dave!
```

We want to call out aspirate here and mention that this it will be folded merged back into aspire by way of the publishers.

Adds WithHttpCommand(), which lets you define a resource command that sends an HTTP request to your app during development. Useful for triggering endpoints like seed or reset from the dashboard.

```csharp
if (builder.Environment.IsDevelopment())
{
    var resetDbKey = Guid.NewGuid().ToString();

    catalogDbApp.WithEnvironment("DatabaseResetKey", resetDbKey)
                .WithHttpCommand("/reset-db", "Reset Database",
                    commandOptions: new()
                    {
                        Description = "Reset the catalog database to its initial state. This will delete and recreate the database.",
                        ConfirmationMessage = "Are you sure you want to reset the catalog database?",
                        IconName = "DatabaseLightning",
                        PrepareRequest = requestContext =>
                        {
                            requestContext.Request.Headers.Add("Authorization", $"Key {resetDbKey}");
                            return Task.CompletedTask;
                        }
                    });
}
```

For more information, see [Custom HTTP commands in .NET Aspire](../fundamentals/http-commands.md).

> ### 🚫 Disable dashboard resource graphs
> 
> The dashboard has the ability to display resource graphs, which show the relationships between resources in your app. This feature is enabled by default, but you can disable it by setting the `Dashboard:UI:DisableResourceGraph` configuration option to `true`.
> 
> For more information, see [.NET Aspire dashboard configuration](../fundamentals/dashboard/configuration.md#other).

Reviewer 4: Delete this, this is not whats new content.
```suggestion
```

## 🚫 Disable dashboard resource graphs

The dashboard has the ability to display resource graphs, which show the relationships between resources in your app. This feature is enabled by default, but you can disable it by setting the `Dashboard:UI:DisableResourceGraph` configuration option to `true`.

For more information, see [.NET Aspire dashboard configuration](../fundamentals/dashboard/configuration.md#other).

## ➕ Adding database resources actually creates a database

There's [plenty of feedback and confusion](https://github.com/dotnet/aspire/issues/7101) around the `AddDatabase` API. The name implies that it adds a database, but it didn't actually create the database. In .NET Aspire 9.2, the `AddDatabase` API now creates a database for the following hosting integrations:

| Hosting integration | API reference |
|--|--|
| [📦 Aspire.Hosting.SqlServer](https://www.nuget.org/packages/Aspire.Hosting.SqlServer) | <xref:Aspire.Hosting.SqlServerBuilderExtensions.AddDatabase*> |
| [📦 Aspire.Hosting.PostgreSql](https://www.nuget.org/packages/Aspire.Hosting.PostgreSql) | <xref:Aspire.Hosting.PostgresBuilderExtensions.AddDatabase*> |

> There are, however; a number of remaining hosting integrations that don't yet create a database:

Reviewer 7: This is a complex sentence that can be simplified rather well. Also, the wording "don't yet" implies a promise that it will, which is more bloggy product talk and not docs. If the plan is to add them, is there a github issue that can be linked to? Adding a statement about the road map would be good after the list.

```suggestion
The following hosting integrations don't support database creation:
```

- [📦 Aspire.Hosting.Milvus](https://www.nuget.org/packages/Aspire.Hosting.Milvus)
- [📦 Aspire.Hosting.MongoDb](https://www.nuget.org/packages/Aspire.Hosting.MongoDb)
- [📦 Aspire.Hosting.MySql](https://www.nuget.org/packages/Aspire.Hosting.MySql)
- [📦 Aspire.Hosting.Oracle](https://www.nuget.org/packages/Aspire.Hosting.Oracle)

> The Azure SQL and Azure PostgreSQL hosting integrations both expose an `AddDatabase` API, but they don't create a database—unless you call their respective `RunAsContainer` methods. For more information, see [Understand Azure integration APIs](../azure/integrations-overview.md#understand-azure-integration-apis).

Reviewer 7: I think this should be moved up, right after the list of SqlServer/PostgreSql packages.

## 🌐 HTTP-based resource command functionality

> [Custom resource commands](../fundamentals/custom-resource-commands.md) now support HTTP-based functionality with the addition of the `WithHttpCommand` API, enabling you to define endpoints for tasks like database migrations or resets. These commands can be executed directly from the .NET Aspire dashboard.

Reviewer 7: Style guide hit
```suggestion
[Custom resource commands](../fundamentals/custom-resource-commands.md) now support HTTP-based functionality with the addition of the `WithHttpCommand` API, enabling you to define endpoints for tasks like database migrations or resets. These commands can be run directly from the .NET Aspire dashboard.
```

}
```

For more information, see [Custom HTTP commands in .NET Aspire](../fundamentals/http-commands.md).

### 🚫 Disable dashboard resource graphs

The dashboard has the ability to display resource graphs, which show the relationships between resources in your app. This feature is enabled by default, but you can disable it by setting the `Dashboard:UI:DisableResourceGraph` configuration option to `true`.

> For more information, see [.NET Aspire dashboard configuration](../fundamentals/dashboard/configuration.md#other).

Reviewer 4: Delete this 😄

**In the API project:**

```csharp

> var builder = Host.CreateApplicationBuilder(args);

Reviewer 4: This is a web application.
